### PR TITLE
SocketServerChannel and EventLoopGroup passed via config instead of Layer Dependency

### LINKED ISSUE #511: Remove layer dependency of EventLoopGroup and ServerChannelFactory from Server

### PR DESCRIPTION
**Pull request by sumawa**
…ngs) instead of dependency layer

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/service/server/TransportFactory.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    def make(trans: Transport) = trans match {
+        case Nio    => nio
+        case Epoll  => epoll
+        case KQueue => kQueue
+        case URing  => uring
+        case Auto   => auto
+    }
+}
+object TransportFactory {
+    def nio: UIO[JChannelFactory[ServerChannel]]    = ChannelFactory.make(() => new NioServerSocketChannel())
```
Comment (kaushik-barodiya) [inline zio-http/src/main/scala/zhttp/service/server/TransportFactory.scala]: Can this be Task instead of UIO?

**Review by kaushik143 — CHANGES_REQUESTED** [review]

File: example/src/main/scala/HelloWorldAdvanced.scala [hunk 2]
```diff
@@ -22,22 +22,22 @@ object HelloWorldAdvanced extends App {
   private val server =
     Server.port(PORT) ++              // Setup port
       Server.paranoidLeakDetection ++ // Paranoid leak detection (affects performance)
-      Server.app(fooBar +++ app)      // Setup the Http app
+      Server.app(fooBar +++ app) ++   // Setup the Http app
+      Server.serverChannel(Transport.Auto)
```
Comment (tusharmath) [inline example/src/main/scala/HelloWorldAdvanced.scala]:
```suggestion
      Server.nio ++ // (Server.epoll, Server.kqueue, Server.uring, Server.auto)
      Server.threads(nThreads) 
```

File: example/src/main/scala/HelloWorldAdvanced.scala [hunk 3]
```diff
@@ -22,12 +20,11 @@ object HelloWorldAdvanced extends App {
   private val server =
     Server.port(PORT) ++              // Setup port
       Server.paranoidLeakDetection ++ // Paranoid leak detection (affects performance)
-      Server.app(fooBar +++ app)      // Setup the Http app
+      Server.app(fooBar +++ app) ++   // Setup the Http app
+      Server.transport(Auto) ++       // (Server.epoll, Server.kqueue, Server.uring, Server.auto)
+      Server.nThreads(1)              // number of threads for event loop
```
Comment (tusharmath) [inline example/src/main/scala/HelloWorldAdvanced.scala]:
```suggestion
      Server.threads(1)              // number of threads for event loop
```

File: zio-http/src/main/scala/zhttp/service/server/TransportFactory.scala [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+import io.netty.channel.epoll.{Epoll, EpollServerSocketChannel}
+import io.netty.channel.kqueue.{KQueue, KQueueServerSocketChannel}
+import io.netty.channel.socket.nio.NioServerSocketChannel
+import io.netty.channel.{ChannelFactory => JChannelFactory, ServerChannel}
+import io.netty.incubator.channel.uring.IOUringServerSocketChannel
+import zhttp.service.{ChannelFactory, EventLoopGroup}
+import zio.{Task, ZManaged}
+
+sealed trait TransportType
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/TransportFactory.scala]:
```suggestion
sealed trait Transport
```

Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/service/server/TransportFactory.scala]:
```suggestion
sealed trait TransportType {
  def eventLoopGroup: ZManaged[Any, Throwable, EventLoopGroup] = ???  
  def channelInitializer: Task[Any, Throwable, JChannelFactory[ServerChannel]] = ???
}
```

**Review by tusharmath — CHANGES_REQUESTED** [review]

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 5]
```diff
@@ -73,33 +84,32 @@ object Server {
...
 
   def make[R](
     server: Server[R, Throwable],
-  ): ZManaged[R with EventLoopGroup with ServerChannelFactory, Throwable, Unit] = {
+  ): ZManaged[R, Throwable, Unit] = {
     val settings = server.settings()
     for {
-      channelFactory <- ZManaged.access[ServerChannelFactory](_.get)
-      eventLoopGroup <- ZManaged.access[EventLoopGroup](_.get)
+      channel        <- ZManaged.fromEffect(Transport.channelInitializer(settings.transport))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
      channelFactory        <- ZManaged.fromEffect(settings.transport.channelFactory)
```

File: example/src/main/scala/example/HelloWorldAdvanced.scala [hunk 6]
```diff
@@ -24,22 +24,22 @@ object HelloWorldAdvanced extends App {
   private val server =
     Server.port(PORT) ++              // Setup port
       Server.paranoidLeakDetection ++ // Paranoid leak detection (affects performance)
-      Server.app(fooBar +++ app)      // Setup the Http app
+      Server.app(fooBar +++ app) ++   // Setup the Http app
+      Server.transport(Auto)          // (Server.epoll, Server.kqueue, Server.uring, Server.auto)
```
Comment (tusharmath) [inline example/src/main/scala/example/HelloWorldAdvanced.scala]:
```suggestion
      Server.transport(Transport.Auto)          // (Server.epoll, Server.kqueue, Server.uring, Server.auto)
```

Alternative API to do something like 

```scala
Server.nio
```

**Comment by tusharmath** [comment]
Closing this because the approach might not work when we want to share the event loop between the client and the server.

**Pull request closed without merging**